adapters/Unity/UnityAdapter/GADMUnityInterstitialAd.m

```diff
@@ -118,95 +118,91 @@ - (void)presentInterstitialFromRootViewController:(UIViewController *)rootViewCo
   id<GADMAdNetworkAdapter> strongAdapter = _adapter;
 
-  if (error == kUnityAdsErrorShowError) {
-    if (strongNetworkConnector && strongAdapter) {
-      [strongNetworkConnector adapterWillDismissInterstitial:strongAdapter];
-      [strongNetworkConnector adapterDidDismissInterstitial:strongAdapter];
-    }
+  if (!strongAdapter) {
+    return;
   }
-}
 
-- (void)unityAdsDidStart:(nonnull NSString *)placementId {
-  // nothing to do
-  // todo: double check if impression need this
-}
+  // The Unity Ads SDK doesn't provide an event for leaving the application, so the adapter assumes
+  // that a click event indicates the user is leaving the application for a browser or deeplink, and
+  // notifies the Google Mobile Ads SDK accordingly.
 
-- (void)unityAdsReady:(nonnull NSString *)placementId {
-  // Logic to mark a placement ready has moved to the UnityAdsLoadDelegate function
-  // unityAdsAdLoaded.
+  [strongNetworkConnector adapterDidGetAdClick:strongAdapter];
+  [strongNetworkConnector adapterWillLeaveApplication:strongAdapter];
 }
 
-#pragma mark - UnityAdsLoadDelegate Methods
-
-- (void)unityAdsAdFailedToLoad:(nonnull NSString *)placementId {
-  dispatch_async(_lockQueue, ^{
-    GADMAdapterUnityMapTableRemoveObjectForKey(_placementInUse, self->_placementID);
-  });
-
-  id<GADMAdNetworkConnector> strongConnector = _connector;
+- (void)unityAdsShowComplete:(NSString *)placementId withFinishState:(UnityAdsShowCompletionState)state {
+  id<GADMAdNetworkConnector> strongNetworkConnector = _connector;
   id<GADMAdNetworkAdapter> strongAdapter = _adapter;
-  if (strongConnector && strongAdapter) {
-    NSString *errorMsg =
-        [NSString stringWithFormat:@"No ad available for the placement ID: %@", _placementID];
-    NSError *error = GADMAdapterUnityErrorWithCodeAndDescription(
-        GADMAdapterUnityErrorPlacementStateNoFill, errorMsg);
-    [strongConnector adapter:strongAdapter didFailAd:error];
+
+  if (!strongAdapter) {
+    return;
   }
+
+  [strongNetworkConnector adapterWillDismissInterstitial:strongAdapter];
+  [strongNetworkConnector adapterDidDismissInterstitial:strongAdapter];
 }
 
-- (void)unityAdsAdLoaded:(nonnull NSString *)placementId {
-  id<GADMAdNetworkConnector> strongNetworkConnector = _connector;
+- (void)unityAdsShowFailed:(NSString *)placementId withError:(UnityAdsShowError)error withMessage:(NSString *)message {
+  id<GADMAdNetworkConnector> strongConnector = _connector;
   id<GADMAdNetworkAdapter> strongAdapter = _adapter;
 
-  if (strongNetworkConnector && strongAdapter) {
-    [strongNetworkConnector adapterDidReceiveInterstitial:strongAdapter];
+  if (!strongAdapter) {
+    return;
   }
+
+  NSError *errorWithDescription =
+      GADMAdapterUnitySDKErrorWithUnityAdsShowErrorAndMessage(error, message);
+  [strongConnector adapter:strongAdapter didFailAd:errorWithDescription];
+
+  [strongConnector adapterWillDismissInterstitial:strongAdapter];
```

Comment:
Per my other comment about moving adapterWillDismissInterstitial: to unityAdsShowStart:, this method should call adapterWillDismissInterstitial: just before adapterWillDismissInterstitial:, to simulate a failure since this interstitial mediation API doesn't have a true "show failure" event.

Reply:
I think you have a typo here, can you clarify?
> this method should call adapterWillDismissInterstitial: just before adapterWillDismissInterstitial

Reply:
Ah, I meant adapterWillPresentInterstitial: just before adapterWillDismissInterstitial:. Since this mediation API doesn't have to indicate "ad failed to show" we simulate that by a quick succession of presented/dismissed.

While you're here, you can also remove the `[strongConnector adapter:strongAdapter didFailAd:errorWithDescription];` call. `didFailAd:` is the load time error, so calling it from Unity's show-time callback will be too late.